### PR TITLE
chore: threat modeling

### PR DESCRIPTION
**Pull request by skyerus**
## This PR

Models OFO's threats.

### Related Issues

#307

### CODE REVIEW

File: docs/threat_model.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Using [OWASP's threat modeling cheat sheet for guidance](https://cheatsheetseries.owasp.org/cheatsheets/Threat_Modeling_Cheat_Sheet.html).
+
+## Application Entry Points
+
+The interfaces through which potential attackers can interact with the application or supply them with data.
+The diagram below models these interfaces, the sections that follow describe them further.
+
+<img src="../images/ofo_threat_model.png" alt="Diagram of ofo threat model">
+
+### Via the application to which ofo sidecars flagd
```
[the remark this answers is not in the document]
Comment (skyerus) [reply docs/threat_model.md]: Good suggestion, thanks, done

File: docs/threat_model.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+Using [OWASP's threat modeling cheat sheet for guidance](https://cheatsheetseries.owasp.org/cheatsheets/Threat_Modeling_Cheat_Sheet.html).
+
+## Application Entry Points
+
+The interfaces through which potential attackers can interact with the application or supply them with data.
+The diagram below models these interfaces, the sections that follow describe them further.
+
+<img src="../images/ofo_threat_model.png" alt="Diagram of OFO threat model">
+
+### Via the application to which OpenFeature Operator (OFO) sidecars flagd
```
Comment (toddbaert) [inline docs/threat_model.md:14]:
```suggestion
### Via the application to which OpenFeature Operator (OFO) injects the flagd sidecar
```

File: docs/threat_model.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+OFO appends flagd as a sidecar container to any pod spec application with valid annotations. The application is then able to evaluate flags by calling flagd (typically via an sdk and flag provider). While a caller of the application could be from a trusted entity (internal infrastructure), it is more prudent to presume the agent to be untrusted.
```
Comment (toddbaert) [inline docs/threat_model.md:16]: Mention the protocol.

File: docs/threat_model.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+## Assets
+
+### Flag source configuration custom resource (CR)
+
+Stored in the kubernetes cluster's etcd instance, this custom resource contains the configuration for the sidecar container (`flagd`) `OFO` appends to the application pod(s).
+
+### Flag configuration CR
+
+Stored in the kubernetes cluster's etcd instance, this custom resource contains the sidecar's (`flagd`) feature flags configuration. It also incorporates configuration for the sidecar container (`flagd`), although, this is deprecated and superseded by the flag source configuration CR.
```
Comment (toddbaert) [inline docs/threat_model.md:31]: I think I would mention the flag configuration rules might contain some identifying information or PII (thinking about targeting rules). They generally should not contain any sort of secret.

File: docs/threat_model.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+Stored in `OFO`'s container's registry, these variables are used to configure the flag source if not declared explicitly by the flag source configuration CR or flag configuration CR.
+
+### flagd flag configuration
+
+`flagd`'s flag configuration is the determinant of the response to any flag request, it is stored in `flagd`'s memory.
+There are numerous ways to set and maintain this state, configured by the flag configuration CR:
+- Within the flag configuration CR 
+- As a file
+- From a remote source
```
Comment (toddbaert) [inline docs/threat_model.md:43]: Maybe link back to the `Flag configuration CR`.

File: docs/threat_model.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+## Controls
+
+### External agency
```
Comment (toddbaert) [inline docs/threat_model.md:47]: What is meant by this? Maybe it's a term I'm unfamiliar with.

File: docs/threat_model.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+#### Cluster state mutation
+
+```mermaid
+flowchart LR
+    user-agent --> kube-apiserver
+```
+
+The user-agent is typically (and therefore assumed to be) internal to the intranet. They make requests to the `kubeapi-server` to mutate the state of the cluster.
```
Comment (toddbaert) [inline docs/threat_model.md:56]: I would add that in this case, the `user-agent` is a cluster admin or developer. `User-agent` is a good term to use, but we should also put it in simpler terms.

File: docs/threat_model.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+2. Mutation of flag configuration CR, fires a [webhook](#webhooks) to `OFO` & is picked up as an event by `flagd` (if configured).
+
+#### Feature flag request
+
+```mermaid
+flowchart LR
+    user-agent --> application --> flagd
+```
+
+The user-agent (e.g. web browser) is typically (and therefore assumed to be) external to the intranet. They make requests to the application to which `OFO` sidecars `flagd`. Internally, the application (typically) uses an `OpenFeature sdk` (dependent on the application's programming language) alongside a `flagd-provider` to make feature flag requests.
```
Comment (toddbaert) [inline docs/threat_model.md:68]:
```suggestion
The user-agent (e.g. web browser) is typically (and therefore assumed to be) external to the intranet. They make requests to the application to which `OFO` sidecars `flagd`. Internally, the application (typically) uses an `OpenFeature SDK` (dependent on the application's programming language) alongside a `flagd-provider` to make feature flag requests.
```

File: docs/threat_model.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+#### flagd state configuration
+
+```mermaid
+flowchart LR
+    flagd --> kube-apiserver
+```
+
+Not applicable if `flagd` is not configured to use the kubernetes sync provider.
+`flagd` makes a request to retrieve its initial flag configuration state from the CR. Subsequently, it registers an event handler to listen to changes to the CR.
```
Comment (toddbaert) [inline docs/threat_model.md:99]: I think we should mention the equivalent filesync provider deployment option, via mounted configMaps

File: docs/threat_model.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+### Via the application to which OpenFeature Operator (OFO) sidecars flagd
+
+OFO appends flagd as a sidecar container to any pod spec application with valid annotations. The application is then able to evaluate flags by calling flagd (typically via an sdk and flag provider). While a caller of the application could be from a trusted entity (internal infrastructure), it is more prudent to presume the agent to be untrusted.
+
+### Via the Kubernetes API server
+
+OFO listens to webhooks from the Kubernetes control plane (its API server), specifically it handles mutations of pods.
+It parses the mutated pod's annotations to determine which CRD to retrieve flagd's flag configuration from, mutating flagd's internal state if necessary.
+
+## Assets
```
Comment (toddbaert) [inline docs/threat_model.md:23]: I think user PII potentially used in evaluation is an asset that should be included here, though in this deployment it never leaves the node or is persisted. I think we should mention that.

**Comment by toddbaert** [comment]
I think this is close to being able to merge.

I would recommend that you add a bit about the filepath sync, and also mention user PII used in evaluation as an asset. The diagram looks really good IMO. I'm not sure if it's worth adding the filesync there or not, it might make it too too busy; I'll leave that to you.

File: docs/threat_model.md (hunk 3 above)
Comment (skyerus) [reply docs/threat_model.md]: Done

File: docs/threat_model.md (hunk 4 above)
Comment (skyerus) [reply docs/threat_model.md:31]: Done

File: docs/threat_model.md (hunk 5 above)
Comment (skyerus) [reply docs/threat_model.md]: Done

File: docs/threat_model.md (hunk 7 above)
Comment (skyerus) [reply docs/threat_model.md]: Good shout, done

File: docs/threat_model.md (hunk 9 above)
Comment (skyerus) [reply docs/threat_model.md]: Done

File: docs/threat_model.md (hunk 10 above)
Comment (skyerus) [reply docs/threat_model.md:23]: Done

File: docs/threat_model.md (hunk 6 above)
Comment (skyerus) [reply docs/threat_model.md]: I've amended it to be `External actions`/`Internal actions` to hopefully be clearer. I'm attempting to convey the difference between an action taken by an agent vs what actions OFO takes internally as a result of them. If you can think of anything more appropriate please let me know

File: docs/threat_model.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+### flagd flag configuration
+
+`flagd`'s flag configuration is the determinant of the response to any flag request, it is stored in `flagd`'s memory.
+There are numerous ways to set and maintain this state, configured by the flag configuration CR:
+- Within the flag configuration CR 
+- As a file
+- From a remote source
+
+## Controls
```
[the remark this answers is not in the document]
Comment (skyerus) [reply docs/threat_model.md:50]: The OWASP cheat sheet declares that it's important to consider [data in transit](https://cheatsheetseries.owasp.org/cheatsheets/Threat_Modeling_Cheat_Sheet.html#consider-data-in-transit-and-data-at-rest), this was my motivation behind modelling how data moves within the system via the actions laid out under the controls section.

Including the mechanisms that are protecting them is a great suggestion. I've updated the controls section to include this.

**Comment by skyerus** [comment]
> I think this is close to being able to merge.
> 
> I would recommend that you add a bit about the filepath sync, and also mention user PII used in evaluation as an asset. I have a question about how you've thought about controls [here](https://github.com/open-feature/open-feature-operator/pull/309#discussion_r1101542806). The diagram looks really good IMO. I'm not sure if it's worth adding the filesync there or not, it might make it too too busy; I'll leave that to you.

I think including the other providers in the diagram will make it overwhelming. I've made a note that the diagram pertains to the kubernetes sync provider.

File: docs/threat_model.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+## Assets
+
+### Flag source configuration custom resource (CR)
+
+Stored in the kubernetes cluster's etcd instance, this custom resource contains the configuration for the sidecar container (`flagd`) `OFO` appends to the application pod(s).
+
+### Flag configuration CR
+
+Stored in the kubernetes cluster's etcd instance, this custom resource contains the sidecar's (`flagd`) feature flags configuration. It also incorporates configuration for the sidecar container (`flagd`), although, this is deprecated and superseded by the flag source configuration CR.
+The feature flags configurations may contain personally identifiable information (used in targeting rules e.g. email domain). They generally should not contain any sort of secret.
```
Comment (toddbaert) [inline docs/threat_model.md:32]: Perfect!

File: docs/threat_model.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+`flagd`'s flag configuration is the determinant of the response to any flag request, it is stored in `flagd`'s memory.
+There are numerous ways to set and maintain this state, configured by the [flag configuration CR](#flag-configuration-cr):
+- Within the flag configuration CR (kubernetes sync provider)
+- As a file (filepath provider)
+- From a remote source (http provider)
+
+### Personally identifiable information
+
+PII may be used in flag evaluation, this data never leaves the node or is persisted.
```
Comment (toddbaert) [inline docs/threat_model.md:48]:
```suggestion
PII may be used in flag evaluation. It is sent from the workload to flagd.
In all deployment patterns currently supported by the operator, this data never leaves the node and is never persisted.
```

**Review by toddbaert — APPROVED** [review]
Approved with one suggestion.

File: docs/threat_model.md (hunk 6 above)
Comment (toddbaert) [reply docs/threat_model.md]: No, this seems clear to me.

**Review by beeme1mr — APPROVED** [review]
Looks good to me. Thanks for putting this together.
